pkg/odo/cli/component/create.go

```diff
@@ -790,16 +792,31 @@ func (co *CreateOptions) devfileRun() (err error) {
 			}
 		} else {
 			// Download devfile from registry
-			params := util.HTTPRequestParams{
-				URL: co.devfileMetadata.devfileRegistry.URL + co.devfileMetadata.devfileLink,
-			}
+			var params util.HTTPRequestParams
 
-			if registryUtil.IsSecure(co.devfileMetadata.devfileRegistry.Name) {
-				token, err := keyring.Get(fmt.Sprintf("%s%s", util.CredentialPrefix, co.devfileMetadata.devfileRegistry.Name), registryUtil.RegistryUser)
+			if strings.Contains(co.devfileMetadata.devfileRegistry.URL, "github") {
+				// Github-based registry
+				params = util.HTTPRequestParams{
+					URL: co.devfileMetadata.devfileRegistry.URL + co.devfileMetadata.devfileLink,
+				}
+				if registryUtil.IsSecure(co.devfileMetadata.devfileRegistry.Name) {
+					token, err := keyring.Get(fmt.Sprintf("%s%s", util.CredentialPrefix, co.devfileMetadata.devfileRegistry.Name), registryUtil.RegistryUser)
+					if err != nil {
+						return errors.Wrap(err, "unable to get secure registry credential from keyring")
+					}
+					params.Token = token
+				}
+			} else {
+				// OCI-based registry
+				u, err := url.Parse(co.devfileMetadata.devfileRegistry.URL)
 				if err != nil {
-					return errors.Wrap(err, "unable to get secure registry credential from keyring")
+					return err
+				}
+				u.Path = path.Join(u.Path, "devfiles", co.devfileMetadata.componentName)
```

Comment:
This could cause problems for stacks with multiple artifacts in the registry, as this will only download the devfile, rather than everything.

Reply:
I think this has been discussed before, for the time being odo wouldn't support download the whole stack. It's consistent with today's Github-based registry, for example the nodejs stack from https://github.com/odo-devfiles/registry/tree/master/devfiles/nodejs has multiple artifacts but if we run `odo create nodejs` it only downloads the devfile.

Reply:
Yes, but part of the reason we're moving to the new registry format is so that those stacks with multiple files can be supported. Shouldn't it be pretty straightforward to change the existing logic just to call `registryLibrary.PullStackFromRegistry(registry, stack)` instead?

CC @kadel

Reply:
Actually it's more complicated than just switch to call another function, currently for OCI-based registry support we use registry REST API instead of registry library (eg. `registryLibrary.PullStackFromRegistry(registry, stack)`), this can align with existing Github-based registry and use the existing odo devfile parser/API (eg. `DownloadFileInMemoryWithCache` which is the wrapper of `HTTPGetRequest`) for component creation. If we want to switch to use registry library we need to do the following work/design:
1. Add additional option to `odo create` for downloading the whole stack
2. Modify the registry library and current odo functions to ensure all component creation features won't be affected, this should at least include:
   2.1. Allow registry library support customized context/root folder
   2.2. Ensure download starter project works well
   2.3. Implement the cache mechanism as we can't reuse the existing odo function `DownloadFileInMemoryWithCache`

But I think the key question is do we want to let odo support downloading the whole stack for now, IMO the Github-based registry and OCI-based registry should be consistent.

Reply:
> I think this has been discussed before, for the time being odo wouldn't support download the whole stack. It's consistent with today's Github-based registry, for example the nodejs stack from https://github.com/odo-devfiles/registry/tree/master/devfiles/nodejs has multiple artifacts but if we run `odo create nodejs` it only downloads the devfile.

The extra files in that nodejs stack are for building the image but they are not required for devfile.yaml to function properly. 
This is not a good example and I don't think that we should store a source like this with devfile in the registry. 

If the stack has multiple files, and those files are required for devfile.yaml to function properly odo will need to download everything.

Reply:
> The extra files in that nodejs stack are for building the image but they are not required for devfile.yaml to function properly.
This is not a good example and I don't think that we should store a source like this with devfile in the registry.

As far as I know, currently all extra files are about outer loop configuration (eg. Dockerfile, k8s manifest), I think the nodejs is the only one has extra files within the stack in our registry for now.

> If the stack has multiple files, and those files are required for devfile.yaml to function properly odo will need to download everything. If we still net

Yes, if files are required for devfile.yaml, currently odo can download all of them from the OCI-based registry with this PR, which is similar to Github-based registry.

Given that, I think for the time being I will make the Github-based registry and OCI-based registry consistent, they can both download the files that required devfile.yaml, and we can extend the functionality once we decide how to host the outer loop files.

Reply:
Based on the cabal call discussion, we will make odo download the whole stack by default when user run `odo create` to create devfile component.